tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,511 @@
+// The returned chunk ref is the reference from where the chunk encoding starts for the chunk.
+func (w *HeadReadWriter) WriteChunk(seriesRef uint64, mint, maxt int64, chk chunkenc.Chunk) (chunkRef uint64, err error) {
+	w.wbufLock.Lock()
+	defer w.wbufLock.Unlock()
+
+	if w.shouldCutSegment(len(chk.Bytes())) {
+		if err := w.cut(); err != nil {
+			return 0, err
+		}
+	}
+
+	binary.BigEndian.PutUint64(w.buf[:], seriesRef)
+	if err := w.write(w.buf[:8]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(w.buf[:], uint64(mint))
+	if err := w.write(w.buf[:8]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(w.buf[:], uint64(maxt))
+	if err := w.write(w.buf[:8]); err != nil {
+		return 0, err
+	}
+
+	// The reference is set to the segment index and the offset where
+	// the data starts for this chunk.
+	//
+	// The upper 4 bytes are for the segment index and
+	// the lower 4 bytes are for the segment offset where to start reading this chunk.
+	chunkRef = w.chunkRef(uint64(w.seq()), uint64(w.n))
+
+	w.buf[0] = byte(chk.Encoding())
+	if err := w.write(w.buf[:1]); err != nil {
+		return 0, err
+	}
+
+	n := binary.PutUvarint(w.buf[:], uint64(len(chk.Bytes())))
+	if err := w.write(w.buf[:n]); err != nil {
+		return 0, err
+	}
+
+	if err := w.write(chk.Bytes()); err != nil {
+		return 0, err
+	}
+
+	return chunkRef, w.wbuf.Flush()
+}
+
+func (w *HeadReadWriter) chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+func (w *HeadReadWriter) shouldCutSegment(chunkLength int) bool {
+	return w.n == 0 || // First segment
+		// TODO: tune this boolean, cutting a segment for only 1 chunk would be inefficient.
+		(time.Since(w.curFileStartTime) > w.segmentTime && w.n > HeadSegmentHeaderSize) || // Time duration reached for the existing file.
+		w.n+int64(chunkLength+27) >= MaxSegmentSize
+}
+
+func (w *HeadReadWriter) seq() int {
+	return w.curFileSequence
+}
+
+func (w *HeadReadWriter) write(b []byte) error {
+	n, err := w.wbuf.Write(b)
+	w.n += int64(n)
+	return err
+}
+
+func (w *HeadReadWriter) cut() (err error) {
+	// Sync current tail to disk and close.
+	if err := w.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(w.dirFile, headChunksFormatV1, false, 0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			f.Close()
+		}
+	}()
+
+	w.n = int64(n)
+	oldSeq := w.curFileSequence
+	w.curFileSequence = seq
+
+	// Write current time in milliseconds for mint and 0 for maxt.
+	w.curFileStartTime = time.Now()
+	binary.BigEndian.PutUint64(w.buf[:], uint64(w.curFileStartTime.UnixNano()/1e6))
+	if _, err := f.Write(w.buf[:8]); err != nil {
+		return err
+	}
+	w.n += 8
+	binary.BigEndian.PutUint64(w.buf[:], 0)
+	if _, err := f.Write(w.buf[:8]); err != nil {
+		return err
+	}
+	w.n += 8
+
+	oldFile := w.curFile
+
+	w.curFile = f
+	if w.wbuf != nil {
+		w.wbuf.Reset(f)
+	} else {
+		w.wbuf = bufio.NewWriterSize(f, 1024)
```

Comment:
This seems like rather a small buffer

Reply:
Buffer is flushed for every chunk write, won't 1024B be enough for 1 chunk?

Reply:
Writing chunk by chunk doesn't sound the most efficient, I'd expect us to buffer up a fair bit more.

Reply:
I considered that, but the reading part gets more complicated and also we have to give a chunk reference from `WriteChunk`. Additionally, if there is a crash while chunks are still in the buffer, we would lose a lot of data with time range depending on the scrape interval.

Reply:
The additional complexity could be taken care of, but the data loss during any crash is my main concern.

Reply:
Wouldn't the data be in the WAL?

Reply:
_Right!_

Reply:
I have added some in-memory buffer support now. Some race needs to be taken care, but I feel this will cause a lot of lock contention between reads and writes.